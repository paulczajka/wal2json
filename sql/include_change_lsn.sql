\set VERBOSITY terse

-- predictability
SET synchronous_commit = on;

DROP TABLE IF EXISTS tbl;
CREATE TABLE tbl (id int);

DROP TABLE IF EXISTS results;
CREATE UNLOGGED TABLE results (data json);

SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');

-- One row should have one record and one end_lsn
INSERT INTO tbl VALUES (1);
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');

select count(*) = 1, count(distinct change->>'change_lsn') = 1
from results,
     json_array_elements(data->'change') change
;
DELETE FROM results;

-- Two rows should have two records and two change_lsns
INSERT INTO tbl VALUES (2);
INSERT INTO tbl VALUES (3);
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');

select count(*) = 2, count(distinct change->>'change_lsn') = 2
from results,
     json_array_elements(data->'change') change
;
DELETE FROM results;

-- Two rows in one transaction should have one record and two change_lsns
INSERT INTO tbl VALUES (4), (5);
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');

select count(distinct data) = 1, count(distinct change->>'change_lsn') = 2
from results,
     json_array_elements(data->'change') change
;
DELETE FROM results;

-- Two rows in one transaction with write in chunks should have two records and two change_lsn
INSERT INTO tbl VALUES (6), (7);
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');

select count(*) = 2, count(distinct change->>'change_lsn') = 2
from results,
     json_array_elements(data->'change') change
;
DELETE FROM results;

-- Two rows in one COPY with should have two records and one change_lsn
COPY tbl FROM STDIN;
8
9
\.
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');

select count(*) = 2, count(distinct change->>'change_lsn') = 1
from results,
     json_array_elements(data->'change') change
;
DELETE FROM results;

SELECT 'stop' FROM pg_drop_replication_slot('regression_slot');

