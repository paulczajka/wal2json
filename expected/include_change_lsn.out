\set VERBOSITY terse
-- predictability
SET synchronous_commit = on;
DROP TABLE IF EXISTS tbl;
NOTICE:  table "tbl" does not exist, skipping
CREATE TABLE tbl (id int);
DROP TABLE IF EXISTS results;
NOTICE:  table "results" does not exist, skipping
CREATE UNLOGGED TABLE results (data json);
SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');
 ?column? 
----------
 init
(1 row)

-- One row should have one record and one end_lsn
INSERT INTO tbl VALUES (1);
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');
select count(*) = 1, count(distinct change->>'change_lsn') = 1
from results,
     json_array_elements(data->'change') change
;
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

DELETE FROM results;
-- Two rows should have two records and two change_lsns
INSERT INTO tbl VALUES (2);
INSERT INTO tbl VALUES (3);
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');
select count(*) = 2, count(distinct change->>'change_lsn') = 2
from results,
     json_array_elements(data->'change') change
;
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

DELETE FROM results;
-- Two rows in one transaction should have one record and two change_lsns
INSERT INTO tbl VALUES (4), (5);
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');
select count(distinct data) = 1, count(distinct change->>'change_lsn') = 2
from results,
     json_array_elements(data->'change') change
;
ERROR:  could not identify an equality operator for type json at character 23
DELETE FROM results;
-- Two rows in one transaction with write in chunks should have two records and two change_lsn
INSERT INTO tbl VALUES (6), (7);
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');
select count(*) = 2, count(distinct change->>'change_lsn') = 2
from results,
     json_array_elements(data->'change') change
;
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

DELETE FROM results;
-- Two rows in one COPY with should have two records and one change_lsn
COPY tbl FROM STDIN;
INSERT INTO results SELECT data::json FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-change-lsn', '1');
select count(*) = 2, count(distinct change->>'change_lsn') = 1
from results,
     json_array_elements(data->'change') change
;
 ?column? | ?column? 
----------+----------
 t        | t
(1 row)

DELETE FROM results;
SELECT 'stop' FROM pg_drop_replication_slot('regression_slot');
 ?column? 
----------
 stop
(1 row)

